stdlib/public/core/StringUTF16View.swift

```diff
@@ -518,6 +519,105 @@ extension _StringGuts {
 }
 
 extension String.UTF16View {
+  
+  @inline(__always)
+  internal func _utf16Length<U: SIMD, S: SIMD>(
+    readPtr: inout UnsafeRawPointer,
+    endPtr: UnsafeRawPointer,
+    unsignedSIMDType: U.Type,
+    signedSIMDType: S.Type
+  ) -> Int where U.Scalar == UInt8, S.Scalar == Int8 {
+    var utf16Count = 0
+    
+    while readPtr + MemoryLayout<U>.stride < endPtr {
+      //Find the number of continuations (0b10xxxxxx)
+      let sValue = Builtin.loadRaw(readPtr._rawValue) as S
+      let continuations = S.zero.replacing(with: S.one, where: sValue .< -65 + 1)
+      let continuationCount = Int(continuations.wrappedSum())
+            
+      //Find the number of 4 byte code points (0b11110xxx)
+      let uValue = Builtin.loadRaw(readPtr._rawValue) as U
+      let fourBytes = U.zero.replacing(with: U.one, where: uValue .>= 0b11110000)
+      let fourByteCount = Int(fourBytes.wrappedSum())
+            
+      utf16Count &+= (U.scalarCount - continuationCount) + fourByteCount
```

Comment:
Horizontal operations (wrappedSum) are expensive; usually in SIMD code we keep the accumulation in vector until forced to map down to a scalar rather, than accumulating every loop iteration. It's a little bit of a pain with Int8 because you can only do 127 accumulations before you have to worry about overflow, but still doing one horizontal operation every N iterations is much better than 2/iteration.

The easy change here would be to do:

`&+= U.scalarCount + (fourByteCount - continuationCount).wrappedSum()`

so that you only do half as many as currently.

Reply:
Sure, I'll make a followup patch

Reply:
(The optimizer might be managing to pull this apart for you already, but it would be nice to make explicit anyway.)

Reply:
https://github.com/apple/swift/pull/41897

Reply:
(I'm hoping once wider vectors are feasible that the common case here, post-breadcrumb offsets, will only ever iterate once because we can just do the entire 64 bytes at once)